presto-main/src/main/java/com/facebook/presto/execution/buffer/PartitionedOutputBuffer.java

```diff
@@ -200,6 +222,30 @@ public void enqueue(int partitionNumber, List<SerializedPage> pages)
         serializedPageReferences.forEach(SerializedPageReference::dereferencePage);
     }
 
+    @Override
+    public void setNoMorePagesForLifespan(Lifespan lifespan)
+    {
+        noMorePagesForLifespan.add(lifespan);
+    }
+
+    @Override
+    public void registerLifespanFinishCallback(Consumer<Lifespan> callback)
+    {
+        checkState(lifespanFinishedCallback == null);
+        this.lifespanFinishedCallback = callback;
+    }
+
+    @Override
+    public boolean isLifespanFinished(Lifespan lifespan)
+    {
+        if (!noMorePagesForLifespan.contains(lifespan)) {
+            return false;
+        }
+
+        AtomicLong outstandingPage = outstandingPagePerLifespan.get(lifespan);
+        return outstandingPage == null || outstandingPage.get() == 0;
```

Comment:
Why not just `return outstandingPagePerLifespan.get(lifespan).get() == 0;` similar to other buffers?

Reply:
See https://github.com/prestodb/presto/pull/12336/commits/549a34ca19644081ac8ce01b073338e1ac2eadfa#r257056902 , sorry for the confusion.